docs/RationaleSeq.md

```diff
@@ -0,0 +1,70 @@
+# Seq(uential) Dialect Rationale
+
+This document describes various design points of the `seq` dialect, why it is
+the way it is, and current status. This follows in the spirit of other [MLIR
+Rationale docs](https://mlir.llvm.org/docs/Rationale/).
+
+## Introduction
+
+Digital logic is generally split into two orthogonal categories:
+combinational and sequential. CIRCT contains a `comb` dialect to model the
+basic combinational operations and the (future) `seq` dialect which is
+discussed here. The intention of the `seq` dialect is to provide a set of
+stateful constructs which can be used to model sequential logic, independent
+of the output method (e.g. SystemVerilog).
+
+## Definitions
+
+For the sake of precision, we use the following definitions:
+
+- **Latch:** A memory element which is only sensitive to the levels of its
+inputs. Has no clock. Example: SR Latch.
+- **Clocked (gated) latch:** A latch wherein the inputs are gated by a clock.
+Transparent the entire time the clock is high. Generally referred to as a
+"latch". Examples: "gated SR latch", "D latch".
+- **Flip-flop:** An edge-sensitive memory element. Captures the input value
+on one or both clock edges. Variants: posedge FF, negedge FF,
+"edge-sensitive" FF (captures the input value on both edges), resettable FF.
+- **Register:** A synchronous, resettable memory element. Can be implemented
+using any of the above "circuit level" elements.
```

Comment:
Just a thought reading this but "one of these things is not like the other".  Register is a much higher abstraction level with a lot of design flexibility and freedom.  I'm not sure that handling it in the same way as the other better-defined and better-constrained elements is the obvious way to go.

Reply:
Another thought: the semantics of a D-latch are fairly distinct from a D-flip-flop, and generally you cannot simply substitute one for the other (since the D-latch becomes transparent, breaking the timing isolation "contract" that flip-flops provide). My intuition would be that you always want to specify whether the storage element you are talking about in your circuit has this timing-arc-breaking behaviour of an FF, or this transparent nature of a latch. Not being able to do so could be too vague to describe anything (e.g., you wouldn't be able to describe a state machine or counter).

Reply:
To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".

You're correct that replacing (edge-triggered) FFs with (clocked) latches would not work in the general case, but with careful timing analysis you can. So you can implement a state machine with latches so long as you have a sufficiently short clock high period, either as a result of a high frequency or duty cycle.

Either way, I think we want to model things at a higher level wherein we don't have to reason about timing then do the timing analysis and device selection at a lower-level. In most cases (I think), one just wants to reason about registers in the way `always_ff` does and let the synthesizer select FF vs latch and worry about the timing. Anything beyond that is an optimization to top of pure digital sequential logic, IMO.

Reply:
> To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".

Just nitpicking: I don't think an SR latch can become transparent (you can set or reset it with the S or R strobe). You probably can do something like that with a gated SR latch when you invert the S or R input and short them together, but then you've essentially built a D Latch ;-) (https://en.wikibooks.org/wiki/Digital_Circuits/Latches#Gated_SR_latch)

> You're correct that replacing (edge-triggered) FFs with (clocked) latches would not work in the general case, but with careful timing analysis you can. So you can implement a state machine with latches so long as you have a sufficiently short clock high period, either as a result of a high frequency or duty cycle.

Fair point! My mind was too fixed on this purely-FF / conservative timing use case.

> Either way, I think we want to model things at a higher level wherein we don't have to reason about timing then do the timing analysis and device selection at a lower-level. In most cases (I think), one just wants to reason about registers which are assumed to break timing arcs and then let the synthesizer select FF vs latch and worry about the timing. Anything beyond that is an optimization on top of pure digital sequential logic, IMO.

Fully agreed, that makes a lot of sense. In that case opting for a latch in the IR would probably stem from the user explicitly asking for one, and in general we'd just go for an abstract register, right?

Reply:
> > To address your two comments Fabian: my understanding is that a gated SR latch and D latch are functionally equivalent (modulo the data signaling) as they both have an enable (strobe) signal which makes them go transparent. A D-FF is technically the same as a d-latch, (modulo the enable vs clock labeling) but most of the time when people say "flip flop", they mean "edge-triggered flip flop".
> 
> Just nitpicking: I don't think an SR latch can become transparent (you can set or reset it with the S or R strobe). You probably can do something like that with a gated SR latch when you invert the S or R input and short them together, but then you've essentially built a D Latch 😉 (https://en.wikibooks.org/wiki/Digital_Circuits/Latches#Gated_SR_latch)

Actually, if you're talking about a dual-rail encoding, then it does become "transparent". 🤪


> Fully agreed, that makes a lot of sense. In that case opting for a latch in the IR would probably stem from the user explicitly asking for one, and in general we'd just go for an abstract register, right?

Yep. Or this reg op could lower to a latch op.